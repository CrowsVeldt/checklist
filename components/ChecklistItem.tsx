import { AntDesign } from "@expo/vector-icons";
import { useState } from "react";
import { Pressable, StyleSheet, View } from "react-native";
import Animated, {
  useAnimatedStyle,
  withTiming,
} from "react-native-reanimated";

export default function ChecklistItem({
  id,
  title,
  required,
  setFinished,
}: {
  id: string;
  title: string;
  required: boolean;
  setFinished: (id: string) => void;
}) {
  const [checked, setChecked] = useState<boolean>(false);

  const animatedStrikethroughStyles = useAnimatedStyle(() => {
    return {
      width: checked ? withTiming("83%") : withTiming("0%"),
      borderWidth: checked ? withTiming(1) : withTiming(0),
      display: checked ? withTiming("flex") : withTiming("none"),
    };
  });

  const animatedTextStyles = useAnimatedStyle(() => {
    return {
      color: checked ? withTiming("gray") : withTiming("black"),
    };
  });

  const animatedViewStyles = useAnimatedStyle(() => {
    return {
      backgroundColor: checked
        ? withTiming("lightgray")
        : withTiming("transparent"),
    };
  });

  return (
    <Animated.View style={[styles.itemContainer, animatedViewStyles]}>
      <Pressable
        onPress={() => {
          setFinished(id);
          setChecked(!checked);
        }}
        style={[styles.checklistItem]}
      >
        <View style={styles.box}>
          {checked && <AntDesign name="check" size={30} color={"black"} />}
        </View>
        <Animated.Text style={animatedTextStyles}>{title}</Animated.Text>
        <Animated.View
          style={[styles.strikeThrough, animatedStrikethroughStyles]}
        />
        {required ? (
          <AntDesign name="exclamationcircleo" size={25} color="black" />
        ) : (
          <AntDesign name="exclamationcircleo" size={25} color="white" />
        )}
      </Pressable>
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  itemContainer: {
    borderEndEndRadius: 25,
    borderEndStartRadius: 25,
    marginBottom: 1,
  },
  checklistItem: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    width: "80%",
  },
  box: {
    borderWidth: 1,
    borderColor: "black",
    height: 30,
    width: 30,
  },
  strikeThrough: {
    height: 1,
    borderColor: "gray",
    alignSelf: "center",
    position: "absolute",
    start: 30,
  },
});
