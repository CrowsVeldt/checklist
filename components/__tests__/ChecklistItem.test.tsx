import { render } from "@testing-library/react-native";
import ChecklistItem from "../ChecklistEntryItem";

describe("<ChecklistItem />", () => {
  test("ChecklistItem renders correctly", () => {
    const tree = render(<ChecklistItem title="title" required={false}/>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
