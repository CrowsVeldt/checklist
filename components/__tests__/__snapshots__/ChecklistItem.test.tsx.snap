// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChecklistItem /> ChecklistItem renders correctly 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "width": "80%",
    }
  }
>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": false,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderColor": "#657786",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 20,
          "width": 20,
        },
        undefined,
        false,
        false,
        undefined,
        false,
      ]
    }
  />
  <Text
    style={{}}
  >
    title
  </Text>
  <Text />
</View>
`;
